File: app/src/main/java/no/nav/sosialhjelp/soknad/domain/SoknadUnderArbeid.java
package no.nav.sosialhjelp.soknad.domain;

import no.nav.sbl.soknadsosialhjelp.soknad.JsonInternalSoknad;

import java.time.LocalDateTime;

import static org.apache.commons.lang3.StringUtils.isEmpty;

public class SoknadUnderArbeid {
    private Long soknadId;
    private Long versjon;
    private String behandlingsId;
    private String tilknyttetBehandlingsId;
    private String eier;
    private JsonInternalSoknad jsonInternalSoknad;
    private SoknadInnsendingStatus innsendingStatus;
    private LocalDateTime opprettetDato;
    private LocalDateTime sistEndretDato;

    public boolean erEttersendelse() {
        return !isEmpty(tilknyttetBehandlingsId);
    }

    public Long getSoknadId() {
        return soknadId;
    }

    public SoknadUnderArbeid withSoknadId(Long soknadId) {
        this.soknadId = soknadId;
        return this;
    }

    public void setSoknadId(Long soknadId) {
        this.soknadId = soknadId;
    }

    public Long getVersjon() {
        return versjon;
    }

    public SoknadUnderArbeid withVersjon(Long versjon) {
        this.versjon = versjon;
        return this;
    }

    public void setVersjon(Long versjon) {
        this.versjon = versjon;
    }

    public String getBehandlingsId() {
        return behandlingsId;
    }

    public SoknadUnderArbeid withBehandlingsId(String behandlingsId) {
        this.behandlingsId = behandlingsId;
        return this;
    }

    public String getTilknyttetBehandlingsId() {
        return tilknyttetBehandlingsId;
    }

    public SoknadUnderArbeid withTilknyttetBehandlingsId(String tilknyttetBehandlingsId) {
        this.tilknyttetBehandlingsId = tilknyttetBehandlingsId;
        return this;
    }

    public String getEier() {
        return eier;
    }

    public SoknadUnderArbeid withEier(String eier) {
        this.eier = eier;
        return this;
    }

    public SoknadInnsendingStatus getInnsendingStatus() {
        return innsendingStatus;
    }

    public SoknadUnderArbeid withInnsendingStatus(SoknadInnsendingStatus innsendingStatus) {
        this.innsendingStatus = innsendingStatus;
        return this;
    }

    public void setInnsendingStatus(SoknadInnsendingStatus innsendingStatus) {
        this.innsendingStatus = innsendingStatus;
    }

    public LocalDateTime getOpprettetDato() {
        return opprettetDato;
    }

    public SoknadUnderArbeid withOpprettetDato(LocalDateTime opprettetDato) {
        this.opprettetDato = opprettetDato;
        return this;
    }

    public LocalDateTime getSistEndretDato() {
        return sistEndretDato;
    }

    public SoknadUnderArbeid withSistEndretDato(LocalDateTime sistEndretDato) {
        this.sistEndretDato = sistEndretDato;
        return this;
    }

    public void setSistEndretDato(LocalDateTime sistEndretDato) {
        this.sistEndretDato = sistEndretDato;
    }

    public JsonInternalSoknad getJsonInternalSoknad() { return jsonInternalSoknad; }

    public SoknadUnderArbeid withJsonInternalSoknad(JsonInternalSoknad jsonInternalSoknad) {
        this.jsonInternalSoknad = jsonInternalSoknad;
        return this;
    }
}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/db/repositories/soknadunderarbeid/SoknadUnderArbeid.kt
package no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid

import no.nav.sbl.soknadsosialhjelp.soknad.JsonInternalSoknad
import no.nav.sbl.soknadsosialhjelp.soknad.personalia.JsonPersonalia
import no.nav.sosialhjelp.soknad.v2.eier.Eier
import no.nav.sosialhjelp.soknad.v2.eier.Kontonummer
import no.nav.sosialhjelp.soknad.v2.navn.Navn
import java.time.LocalDateTime
import java.util.*

data class SoknadUnderArbeid(
    var soknadId: Long = 0L, // dummy verdi, settes i DB
    var versjon: Long,
    var behandlingsId: String,
    private val tilknyttetBehandlingsId: String? = null,
    var eier: String,
    var jsonInternalSoknad: JsonInternalSoknad?,
    var status: SoknadUnderArbeidStatus,
    var opprettetDato: LocalDateTime,
    var sistEndretDato: LocalDateTime
)

enum class SoknadUnderArbeidStatus {
    UNDER_ARBEID, LAAST
}

fun SoknadUnderArbeid.toV2Eier(): Eier? {
    return jsonInternalSoknad?.soknad?.data?.personalia?.toV2Eier(UUID.fromString(this.behandlingsId))
}

private fun JsonPersonalia.toV2Eier(soknadId: UUID): Eier {
    return Eier(
        soknadId = soknadId,
        statsborgerskap = statsborgerskap?.verdi,
        nordiskBorger = nordiskBorger?.verdi,
        kontonummer = kontonummer?.let { Kontonummer(fraRegister = it.verdi) },
        navn = Navn(
            fornavn = navn.fornavn,
            mellomnavn = navn.mellomnavn,
            etternavn = navn.etternavn
        )
    )
}
